sample-code/spring-app/src/main/java/com/sap/ai/sdk/app/services/OpenAiServiceV2.java

```diff
@@ -102,64 +92,36 @@ public OpenAiChatCompletionResponse chatCompletionImage(@Nonnull final String li
   @Nonnull
   public OpenAiChatCompletionResponse chatCompletionToolExecution(
       @Nonnull final String location, @Nonnull final String unit) {
-
-    // 1. Define the function
-    final Map<String, Object> schemaMap = generateSchema(WeatherMethod.Request.class);
-    final var function =
-        new FunctionObject()
-            .name("weather")
-            .description("Get the weather for the given location")
-            .parameters(schemaMap);
-    final var tool = new ChatCompletionTool().type(FUNCTION).function(function);
-
     final var messages = new ArrayList<OpenAiMessage>();
     messages.add(OpenAiMessage.user("What's the weather in %s in %s?".formatted(location, unit)));
 
-    // Assistant will call the function
-    final var request = new OpenAiChatCompletionRequest(messages).withTools(List.of(tool));
+    // 1. Define the function
+    final var weatherFunction =
+        new OpenAiFunctionTool("weather", WeatherMethod.Request.class)
+            .withDescription("Get the weather for the given location");
+
+    // 2. Assistant calls the function
+    final var request =
+        new OpenAiChatCompletionRequest(messages).withOpenAiTools(List.of(weatherFunction));
     final OpenAiChatCompletionResponse response =
         OpenAiClient.forModel(GPT_4O_MINI).chatCompletion(request);
-
-    // 2. Optionally, execute the function.
     final OpenAiAssistantMessage assistantMessage = response.getMessage();
-    messages.add(assistantMessage);
 
+    // 3. Execute the function
     final OpenAiToolCall toolCall = assistantMessage.toolCalls().get(0);
     if (!(toolCall instanceof OpenAiFunctionCall functionCall)) {
       throw new IllegalArgumentException(
           "Expected a function call, but got: %s".formatted(assistantMessage));
     }
+    final WeatherMethod.Request arguments = functionCall.getArgumentsAsObject(weatherFunction);
+    final WeatherMethod.Response currentWeather = WeatherMethod.getCurrentWeather(arguments);
 
-    final WeatherMethod.Request arguments =
-        parseJson(functionCall.getArguments(), WeatherMethod.Request.class);
-    final WeatherMethod.Response weatherMethod = WeatherMethod.getCurrentWeather(arguments);
-
-    messages.add(OpenAiMessage.tool(weatherMethod.toString(), functionCall.getId()));
-
-    // Send back the results, and the model will incorporate them into its final response.
+    // 4. Send back the results, and the model will incorporate them into its final response.
+    messages.add(assistantMessage);
+    messages.add(OpenAiMessage.tool(currentWeather.toString(), functionCall.getId()));
```

Comment:
`OpenAiMessage.tool(currentWeather.toString(), functionCall.getId())` could also be a convenience so th customer doesn't have to build it.